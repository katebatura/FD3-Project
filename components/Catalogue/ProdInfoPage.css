.productPage-container {
    display: flex;
    justify-content: start;
    color: black;
}
.productPage-title {
    width: 100%;
    color: black;
}

.productPage-img {
    width: 200px;
    margin-right: 40px;
}
.productPage-price {
    color: #19bde8;
    font-weight: bold;
    font-size: 24px;
}