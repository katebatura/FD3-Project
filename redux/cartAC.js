const QTY_ADD = 'QTY_ADD';
const PROD_ADD = 'PROD_ADD';
const PROD_DEL = 'PROD_DEL';
const SUM_QTY = 'SUM_QTY';
const DELIVERY_CHNG = 'DELIVERY_CHNG';

const prod_add = function(prodid, info) {
  return {
    type: PROD_ADD,
    prodid: prodid,
    info: info,
  };
}

const prod_del = function(prodid) {
  return {
    type: PROD_DEL,
    prodid: prodid,
  };
}

const qty_add = function(prodid, newQTY, newSum) {
  return {
    type: QTY_ADD,
    prodid: prodid,
    newQTY: newQTY,
    newSum: newSum
  };
}

const sum_qty = function(qty) {
  return {
    type: SUM_QTY,
    qty: qty,
  }
}

const delivery_chng = function(num) {
  return {
    type: DELIVERY_CHNG,
    num: num,
  }
}

export {
  prod_add, PROD_ADD,
  prod_del, PROD_DEL,
  qty_add, QTY_ADD,
  sum_qty, SUM_QTY,
  delivery_chng, DELIVERY_CHNG,
}
