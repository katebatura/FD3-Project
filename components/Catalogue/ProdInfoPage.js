"use strict";

import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';

import {connect} from 'react-redux';
import { prod_add, sum_qty } from '../../redux/cartAC';

import { setLocalStorage } from '../../LocalStorage/LocalStorage';

import './Product.css';

class ProdInfoPage extends React.PureComponent {
  
  static propTypes = { 
    info: PropTypes.shape({
        id: PropTypes.number.isRequired,
        img: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired,
        category: PropTypes.string.isRequired,
      }),
  };

  state = {
    info: this.props.info,
  };

  
  newQTYRef = null;

  setNewQTYRef = (ref) => {
    this.newQTYRef = ref
  }

  addIntoCart = () => {
    let qty = +this.newQTYRef.value;
    let sum = qty * this.props.info.price;
    let info = {...this.state.info, qty: qty, sum: sum};

    this.props.dispatch( prod_add(this.state.info.id, info) ); //добавляем в корзину
    this.props.dispatch( sum_qty(qty) ); //добавляем количество продуктов в корзине для отображения в header

    setLocalStorage(this.state.info.id, info);
  }
  
  render() {

    return (
        <div className="product-container">
          <h3>{this.props.info.name}</h3>
          <img src = {this.props.info.img} className = "product-img" />
          <p>{this.props.info.price}</p>
          <input type='number' step = "1" min = "0" defaultValue = {1} title = "Кол-во" ref = {this.setNewQTYRef} />
          <input type = "button" value = "add" onClick = {this.addIntoCart} />
        </div>             
    );

  }

}

 
const mapStateToProps = function (state) {
  // этому компоненту ничего не нужно из хранилища Redux
  return { }; 
};

export default connect(mapStateToProps)(ProdInfoPage);
