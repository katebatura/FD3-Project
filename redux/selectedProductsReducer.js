import { SUM_QTY } from './cartAC';

const initState = {
    
    productsQTY: 0,
  
  }
  
  
  function selectedProductsReducer(state = initState, action) {
    switch (action.type) {

      case SUM_QTY: { 
        let newQTY = state.productsQTY + action.qty;
        let newState = {...state,
          productsQTY:  newQTY};
        return newState;
      }

      default:
        return state;
    }
  }
  
  export default selectedProductsReducer;