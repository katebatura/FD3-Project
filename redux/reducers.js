import { combineReducers } from 'redux';

import productsListReducer from "./productsListReducer";
import cartReducer from "./cartReducer";
import selectedProductsReducer from "./selectedProductsReducer";

let combinedReducer = combineReducers({
    // редьюсер countersReducer отвечает за раздел state под именем counters
    products: productsListReducer,
    cart: cartReducer, 
    selectedProducts: selectedProductsReducer,
});

export default combinedReducer;
