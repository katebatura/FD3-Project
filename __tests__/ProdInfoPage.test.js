"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';

import { createStore } from 'redux';
import combinedReducer from '../redux/reducers.js';
import { Provider } from 'react-redux';

import ProdInfoPage from '../components/Catalogue/ProdInfoPage';

const prod = {
  "id": 1,
  "name": "AZOLLA ZS 32",
  "brand": "TOTAL",
  "category": "Индустриальные масла",
  "price": 1052.94,
  "description": "Гидравлическое масло",
  "imgUrl": "./img/bochka.jpg",
  "qty": 1,
  "sum": 1052.94
};
    

let store=createStore(combinedReducer);

test('внешний вид и работа ProdInfoPage', () => {

  // создаём тестовую версию компонента
  const component = renderer.create(
    <Router>
      <Provider store={store}>
        <ProdInfoPage info = {prod} />
      </Provider>
    </Router>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // найдём в вёрстке компонента саму кнопку
  const buttonAdd = component.root.find( el => el.type=='button' /*&& el.props.aaa == 'bbb'*/ ); 
  // и "нажмём" на неё
  buttonAdd.props.onClick();
  
  // получаем уже изменённый снэпшот
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // "нажмём" кнопку ещё раз
  buttonAdd.props.onClick();
  
  // и получаем окончательный снэпшот
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

        
});
