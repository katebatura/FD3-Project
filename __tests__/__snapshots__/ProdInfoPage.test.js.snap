// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`внешний вид и работа ProdInfoPage 1`] = `
<div>
  <div
    className="breadcrumbs-container"
  >
    <a
      aria-current={null}
      className="breadcrumbs"
      href="/"
      onClick={[Function]}
      style={undefined}
    >
      Главная 
    </a>
    <span
      className="breadcrumbs-arr"
    >
       → 
    </span>
    <a
      aria-current={null}
      className="breadcrumbs"
      href="/catalogue"
      onClick={[Function]}
      style={undefined}
    >
      Каталог
    </a>
    <span
      className="breadcrumbs-arr"
    >
       → 
    </span>
    <a
      aria-current={null}
      className="breadcrumbs"
      href="/catalogue/Индустриальные масла"
      onClick={[Function]}
      style={undefined}
    >
      Индустриальные масла
    </a>
    <span
      className="breadcrumbs-arr"
    >
       → 
    </span>
    <a
      aria-current={null}
      className="breadcrumbs"
      href="/catalogue/Индустриальные масла/1"
      onClick={[Function]}
      style={undefined}
    >
      AZOLLA ZS 32
    </a>
  </div>
  <hr />
  <h2
    className="productPage-title"
  >
    AZOLLA ZS 32
  </h2>
  <div
    className="productPage-container"
  >
    <div
      className="productPage-img-container"
    >
      <img
        className="productPage-img"
        src="./img/bochka.jpg"
      />
    </div>
    <div
      className="productPage-info"
    >
      <p>
        Индустриальные масла
      </p>
      <p>
        TOTAL
      </p>
      <p>
        Гидравлическое масло
      </p>
      <p
        className="productPage-price"
      >
        1052.94 руб.
      </p>
      <div
        className="addCart-block"
      >
        <input
          className="addQty_button"
          defaultValue={1}
          min="0"
          step="1"
          title="Кол-во"
          type="number"
        />
        <button
          className="addCart_button"
          onClick={[Function]}
        >
          <i
            className="fas fa-cart-arrow-down"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
